### PR TITLE
(8.0) PS-8879: OOM when alter column to compression

### PR DESCRIPTION
**Pull request by kamil-holubicki**
https://jira.percona.com/browse/PS-8879

Problem:
If the large table containing BLOB column is altered in the way that
the COLUMN_FORMAT becomes COMPRESSED
(like ALTER TABLE t1 MODIFY a10 BLOB COLUMN_FORMAT COMPRESSED),
we observe huge memory utilization during the ALTER. This can lead to
OOM situation when mysqld gets killed by the system.

Cause:
Fix for PS-7657 (https://github.com/percona/percona-server/commit/a943a8fedf5a863f32735603815490733d9c2783) removed emptying of compress_heap from before
insertion of the row into InnoDB. This was because during the UPDATE,
prior to storing the row, the compressed column is read from InnoDB
and decompressed. Its decompressed version is kept on
prebuilt->compress_heap. Then when we store the row, we need
the compress_heap to be still valid to have access to the BLOB column
value.
However, if we don't clean the compress_heap, it constantly grows in
cases like ALTER ... COLUMN_FORMAT COMPRESSED, because during this
process we copy all data from altered table to temp table in a loop.

Solution:
Rework the original fix in the way that compress_heap is emptied after
the row is inserted.

### CODE REVIEW

**Comment by kamil-holubicki** [comment]
https://ps80.cd.percona.com/view/8.0%20parallel%20MTR/job/percona-server-8.0-pipeline-parallel-mtr/357

File: storage/innobase/row/row0mysql.cc [hunk 1]
```diff
@@ -2176,6 +2177,12 @@ static dberr_t row_insert_for_mysql_using_ins_graph(const byte *mysql_rec,
 @param[in,out]  prebuilt        prebuilt struct in MySQL handle
 @return error code or DB_SUCCESS*/
 dberr_t row_insert_for_mysql(const byte *mysql_rec, row_prebuilt_t *prebuilt) {
+  raii::Sentry<> compress_heap_cleaner([prebuilt] {
```
Comment (satya-bodapati) [inline storage/innobase/row/row0mysql.cc:2192]: Can you please put a comment on why you are freeing compress_heap after insert?

**Review by satya-bodapati — APPROVED** [review]
rest is all OK. We will deal partition problem separately ;)

File: storage/innobase/row/row0mysql.cc (hunk 1 above)
Comment (kamil-holubicki) [reply storage/innobase/row/row0mysql.cc:2192]: Here you go.... :)

**Pull request merged**